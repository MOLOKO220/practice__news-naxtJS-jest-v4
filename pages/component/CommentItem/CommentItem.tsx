import React, { useEffect, useState } from "react";

interface commentProps {
  id: number;
}

interface commentData {
  by?: string;
  text?: string;
  time?: number;
}

export default function CommentItem(props: commentProps) {
  // hooks
  const [comment, setComment] = useState<commentData>({});
  const [date, setDate] = useState<any>();

  // get a comment
  useEffect(() => {
    fetch(
      `https://hacker-news.firebaseio.com/v0/item/${props.id}.json?print=pretty`
    )
      .then((response) => {
        return response.json();
      })
      .then((data) => {
        console.log(data);
        setComment(data);
      });
  }, []);

  // change format date
  useEffect(() => {
    const date = new Date(Number(`${comment.time}000`));
    setDate(
      `${date.getFullYear()}.${
        date.getMonth() + 1 < 10
          ? `0${date.getMonth() + 1}`
          : date.getMonth() + 1
      }.${
        date.getDate() < 10 ? `0${date.getDate()}` : date.getDate()
      }  ${date.getHours()}:${date.getMinutes()}`
    );
  }, [comment]);

  return (
    <li className="CommentItem">
      <header>
        <h6>{comment.by}</h6>
        <p>{date}</p>
      </header>
      <p>{comment.text}</p>
    </li>
  );
}
