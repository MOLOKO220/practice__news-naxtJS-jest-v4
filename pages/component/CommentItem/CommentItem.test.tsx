import { render, screen, act } from "@testing-library/react";
import CommentItem from "./CommentItem";

global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () =>
      Promise.resolve({
        by: "danieg",
        descendants: 101,
        id: 35627107,
        kids: [35628027, 35627475, 35627904, 35627413],
        score: 212,
        time: 1681905950,
        title:
          "Making a Linux home server sleep on idle and wake on demand â€“ the simple way",
        type: "story",
        url: "https://dgross.ca/blog/linux-home-server-auto-sleep/",
      }),
  })
) as jest.Mock;

describe("CommentItem", () => {
  it("Fetch", async () => {
    await act(() => render(<CommentItem id={343423534} />));

    expect(fetch).toHaveBeenCalledTimes(1);
    expect(screen.getByText(/danieg/i)).toBeInTheDocument();
  });

  it("snapshot", () => {
    const { container } = render(<CommentItem id={343423534} />);
    expect(container).toMatchSnapshot();
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });
});
