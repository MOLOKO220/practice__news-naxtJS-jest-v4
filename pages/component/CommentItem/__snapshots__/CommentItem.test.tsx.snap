// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentItem snapshot 1`] = `
<div>
  <li
    class="CommentItem"
  >
    <header>
      <h6 />
      <p>
        NaN.NaN.NaN  NaN:NaN
      </p>
    </header>
    <p />
  </li>
</div>
`;
